File: PreStage.Core/Services/XmpService.cs
using System.Xml.Linq;
using PreStage.Core.Models;

namespace PreStage.Core.Services;

public class XmpService
{
    private static readonly XNamespace XapNs = "http://ns.adobe.com/xap/1.0/";
    private static readonly XNamespace DcNs = "http://purl.org/dc/elements/1.1/";
    private static readonly XNamespace RdfNs = "http://www.w3.org/1999/02/22-rdf-syntax-ns#";
    private static readonly XNamespace PhotoshopNs = "http://ns.adobe.com/photoshop/1.0/";
    private static readonly XNamespace PreStageNs = "https://local.codex/prestage/1.0/";

    public XmpData Read(string sidecarPath)
    {
        try
        {
            var doc = XDocument.Load(sidecarPath);
            var rdf = doc.Root?.Element(RdfNs + "RDF");
            if (rdf == null) return new XmpData();

            var desc = rdf.Element(RdfNs + "Description");
            if (desc == null) return new XmpData();

            var data = new XmpData();
            data.RawDocument = doc;

            foreach (var node in desc.DescendantsAndSelf())
                CollectValues(node, data);

            return data;
        }
        catch
        {
            return new XmpData();
        }
    }

    public void Write(string sidecarPath, XmpData data)
    {
        data = NormalizeForWrite(data);
        XDocument doc;

        if (data.RawDocument != null)
        {
            doc = new XDocument(data.RawDocument);
            var rdf = doc.Root?.Element(RdfNs + "RDF");
            var desc = rdf?.Element(RdfNs + "Description");
            if (desc == null) return;

            ApplyPreStageValues(desc, data);

            if (!string.IsNullOrWhiteSpace(data.CreatorTool))
            {
                SetAttributeIfMissing(desc, XapNs + "CreatorTool", data.CreatorTool);
            }
        }
        else
        {
            doc = CreateNewDocument(data);
        }

        doc.Save(sidecarPath);
    }

    private static XDocument CreateNewDocument(XmpData data)
    {
        XNamespace x = "adobe:ns:meta/";
        var doc = new XDocument(
            new XElement(x + "xmpmeta",
                new XAttribute(XNamespace.Xmlns + "x", x),
                new XElement(RdfNs + "RDF",
                    new XAttribute(XNamespace.Xmlns + "rdf", RdfNs),
                    new XElement(RdfNs + "Description",
                    new XAttribute(RdfNs + "about", ""),
                        new XAttribute(XNamespace.Xmlns + "dc", DcNs),
                        new XAttribute(XNamespace.Xmlns + "xap", XapNs),
                        new XAttribute(XNamespace.Xmlns + "photoshop", PhotoshopNs),
                        new XAttribute(XNamespace.Xmlns + "prestage", PreStageNs),
                        !string.IsNullOrWhiteSpace(data.CreatorTool)
                            ? new XAttribute(XapNs + "CreatorTool", data.CreatorTool) : null,
                        new XAttribute(XapNs + "Rating", data.Rating),
                        new XAttribute(XapNs + "Label", data.Label ?? ""),
                        new XAttribute(PhotoshopNs + "Urgency", PhotoshopUrgency(data.Label)),
                        new XAttribute(PreStageNs + "PickState", data.PickState.ToString())
                    )
                )
            )
        );
        return doc;
    }

    private static void CollectValues(XElement element, XmpData data)
    {
        foreach (var attribute in element.Attributes())
        {
            ApplyValue(LocalName(attribute.Name), attribute.Value, data);
        }

        if (!element.HasElements && !string.IsNullOrWhiteSpace(element.Value))
        {
            ApplyValue(LocalName(element.Name), element.Value, data);
        }
    }

    private static void ApplyValue(string name, string? value, XmpData data)
    {
        if (string.IsNullOrWhiteSpace(value))
            return;

        switch (name)
        {
            case "Rating":
                if (int.TryParse(value.Trim(), out var rating))
                {
                    if (rating == -1)
                    {
                        data.Rating = 0;
                        data.PickState = PickState.Rejected;
                    }
                    else
                    {
                        data.Rating = rating;
                    }
                }
                break;
            case "Label":
                data.Label = value;
                break;
            case "CreatorTool":
                data.CreatorTool = value;
                break;
            case "format":
                data.Format = value;
                break;
            case "PickState":
                if (Enum.TryParse<PickState>(value, ignoreCase: true, out var pickState))
                {
                    if (data.PickState == PickState.Rejected && pickState == PickState.Unmarked)
                        break;

                    data.PickState = pickState;
                }
                break;
        }
    }

    private static XmpData NormalizeForWrite(XmpData data)
    {
        if (data.Rating != -1)
            return data;

        return new XmpData
        {
            Rating = 0,
            Label = data.Label,
            CreatorTool = data.CreatorTool,
            Format = data.Format,
            PickState = PickState.Rejected,
            RawDocument = data.RawDocument
        };
    }

    private static void ApplyPreStageValues(XElement desc, XmpData data)
    {
        EnsureNamespace(desc, "xap", XapNs);
        EnsureNamespace(desc, "photoshop", PhotoshopNs);
        EnsureNamespace(desc, "prestage", PreStageNs);

        SetAttribute(desc, XapNs + "Rating", data.Rating.ToString());
        SetAttribute(desc, XapNs + "Label", data.Label ?? "");
        SetAttribute(desc, PhotoshopNs + "Urgency", PhotoshopUrgency(data.Label).ToString());
        SetAttribute(desc, PreStageNs + "PickState", data.PickState.ToString());

        RemoveLegacyElement(desc, XapNs + "Rating");
        RemoveLegacyElement(desc, XapNs + "Label");
    }

    private static void EnsureNamespace(XElement element, string prefix, XNamespace ns)
    {
        var attr = XNamespace.Xmlns + prefix;
        if (element.Attribute(attr) == null)
            element.SetAttributeValue(attr, ns.NamespaceName);
    }

    private static void SetAttribute(XElement element, XName name, string value)
    {
        element.SetAttributeValue(name, value);
    }

    private static void SetAttributeIfMissing(XElement element, XName name, string value)
    {
        if (element.Attribute(name) == null)
            element.SetAttributeValue(name, value);
    }

    private static void RemoveLegacyElement(XElement element, XName name)
    {
        element.Element(name)?.Remove();
    }

    private static string LocalName(XName name) => name.LocalName;

    private static int PhotoshopUrgency(string? label)
    {
        return label?.Trim().ToLowerInvariant() switch
        {
            "red" => 1,
            "yellow" => 2,
            "green" => 3,
            "blue" => 4,
            "purple" => 5,
            _ => 0
        };
    }

    public static string GetSidecarPath(string mediaPath)
    {
        var dir = Path.GetDirectoryName(mediaPath) ?? "";
        var name = Path.GetFileNameWithoutExtension(mediaPath);
        return Path.Combine(dir, name + ".xmp");
    }
}

public class XmpData
{
    public int Rating { get; set; }
    public string? Label { get; set; }
    public string? CreatorTool { get; set; }
    public string? Format { get; set; }
    public PickState PickState { get; set; } = PickState.Unmarked;
    public XDocument? RawDocument { get; set; }
}

File: PreStage-Mac/Sources/PreStage/Services/XMPService.swift
import Foundation

struct XMPValues {
    var rating: Int?
    var colorLabel: ColorLabel?
    var pickState: PickState?
}

enum XMPServiceError: LocalizedError {
    case cannotParseExistingSidecar(URL)
    case missingWritableDescription(URL)

    var errorDescription: String? {
        switch self {
        case .cannotParseExistingSidecar(let url):
            "Could not parse existing XMP sidecar at \(url.path)."
        case .missingWritableDescription(let url):
            "Could not find a writable XMP description in \(url.path)."
        }
    }
}

struct XMPService {
    func applySidecarMetadata(to item: MediaItem) -> MediaItem {
        var updated = item
        let sidecarURL = xmpSidecarURL(for: item.url)
        guard FileManager.default.fileExists(atPath: sidecarURL.path) else {
            return updated
        }
        guard let values = readSidecar(for: item.url) else {
            updated.xmpStatus = .conflict
            return updated
        }

        if let rating = values.rating {
            updated.rating = max(0, min(5, rating))
        }
        if let colorLabel = values.colorLabel {
            updated.colorLabel = colorLabel
        }
        if let pickState = values.pickState {
            updated.pickState = pickState
        }
        updated.xmpStatus = .sidecarFound
        return updated
    }

    func writeSidecar(for item: MediaItem) throws {
        let sidecarURL = xmpSidecarURL(for: item.url)
        if FileManager.default.fileExists(atPath: sidecarURL.path) {
            guard let document = try? XMLDocument(contentsOf: sidecarURL) else {
                throw XMPServiceError.cannotParseExistingSidecar(sidecarURL)
            }
            guard let description = firstDescriptionElement(in: document.rootElement()) else {
                throw XMPServiceError.missingWritableDescription(sidecarURL)
            }
            applyPreStageAttributes(to: description, item: item, preserveExistingCreatorTool: true)
            let data = document.xmlData(options: [.nodePrettyPrint])
            try data.write(to: sidecarURL, options: .atomic)
            return
        }

        try newSidecarXML(for: item).write(to: sidecarURL, atomically: true, encoding: .utf8)
    }

    private func newSidecarXML(for item: MediaItem) -> String {
        let label = item.colorLabel?.xmpLabelName ?? ""
        let pickState = item.pickState.rawValue
        let escapedFilename = item.filename.xmlEscaped
        return """
        <?xpacket begin="" id="W5M0MpCehiHzreSzNTczkc9d"?>
        <x:xmpmeta xmlns:x="adobe:ns:meta/">
          <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
            <rdf:Description rdf:about="\(escapedFilename)"
              xmlns:xmp="http://ns.adobe.com/xap/1.0/"
              xmlns:photoshop="http://ns.adobe.com/photoshop/1.0/"
              xmlns:prestage="https://local.codex/prestage/1.0/"
              xmp:CreatorTool="PreStage"
              xmp:Rating="\(item.rating)"
              xmp:Label="\(label.xmlEscaped)"
              photoshop:Urgency="\(photoshopUrgency(for: item.colorLabel))"
              prestage:PickState="\(pickState)" />
          </rdf:RDF>
        </x:xmpmeta>
        <?xpacket end="w"?>
        """
    }

    private func readSidecar(for mediaURL: URL) -> XMPValues? {
        let sidecarURL = xmpSidecarURL(for: mediaURL)
        guard let document = try? XMLDocument(contentsOf: sidecarURL) else { return nil }
        var values = XMPValues()
        collectValues(from: document.rootElement(), values: &values)
        return values
    }

    private func collectValues(from element: XMLElement?, values: inout XMPValues) {
        guard let element else { return }
        for attribute in element.attributes ?? [] {
            switch normalizedName(attribute.name) {
            case "Rating":
                applyRatingValue(attribute.stringValue, to: &values)
            case "Label":
                values.colorLabel = ColorLabel(displayName: attribute.stringValue ?? "")
            case "PickState":
                values.pickState = PickState(rawValue: attribute.stringValue ?? "")
            default:
                break
            }
        }

        if isLeafTextElement(element), let text = element.stringValue {
            switch normalizedName(element.name ?? element.localName) {
            case "Rating":
                applyRatingValue(text, to: &values)
            case "Label":
                values.colorLabel = ColorLabel(displayName: text)
            case "PickState":
                values.pickState = PickState(rawValue: text)
            default:
                break
            }
        }

        for child in element.children ?? [] {
            collectValues(from: child as? XMLElement, values: &values)
        }
    }

    private func normalizedName(_ rawName: String?) -> String {
        guard let rawName else { return "" }
        let local = rawName.split(separator: ":").last.map(String.init) ?? rawName
        switch local {
        case "Rating":
            return "Rating"
        case "Label":
            return "Label"
        case "PickState":
            return "PickState"
        default:
            return local
        }
    }

    private func isLeafTextElement(_ element: XMLElement) -> Bool {
        guard let children = element.children, !children.isEmpty else { return false }
        return !children.contains { $0 is XMLElement }
    }

    private func applyRatingValue(_ rawValue: String?, to values: inout XMPValues) {
        guard let rating = Int((rawValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)) else { return }
        if rating == -1 {
            values.rating = 0
            if values.pickState == nil {
                values.pickState = .rejected
            }
        } else {
            values.rating = rating
        }
    }

    private func firstDescriptionElement(in element: XMLElement?) -> XMLElement? {
        guard let element else { return nil }
        if element.name == "rdf:Description" || element.localName == "Description" || element.name == "Description" {
            return element
        }

        for child in element.children ?? [] {
            if let match = firstDescriptionElement(in: child as? XMLElement) {
                return match
            }
        }
        return nil
    }

    private func applyPreStageAttributes(to element: XMLElement, item: MediaItem, preserveExistingCreatorTool: Bool) {
        setAttribute("xmlns:xmp", value: "http://ns.adobe.com/xap/1.0/", on: element)
        setAttribute("xmlns:photoshop", value: "http://ns.adobe.com/photoshop/1.0/", on: element)
        setAttribute("xmlns:prestage", value: "https://local.codex/prestage/1.0/", on: element)
        if !preserveExistingCreatorTool || element.attribute(forName: "xmp:CreatorTool") == nil {
            setAttribute("xmp:CreatorTool", value: "PreStage", on: element)
        }
        setAttribute("xmp:Rating", value: "\(item.rating)", on: element)
        setAttribute("xmp:Label", value: item.colorLabel?.xmpLabelName ?? "", on: element)
        setAttribute("photoshop:Urgency", value: "\(photoshopUrgency(for: item.colorLabel))", on: element)
        setAttribute("prestage:PickState", value: item.pickState.rawValue, on: element)
        element.removeAttribute(forName: "photocopy:PickState")
    }

    private func setAttribute(_ name: String, value: String, on element: XMLElement) {
        if let attribute = element.attribute(forName: name) {
            attribute.stringValue = value
        } else {
            element.addAttribute(XMLNode.attribute(withName: name, stringValue: value) as! XMLNode)
        }
    }

    private func xmpSidecarURL(for mediaURL: URL) -> URL {
        mediaURL.deletingPathExtension().appendingPathExtension("xmp")
    }

    private func photoshopUrgency(for label: ColorLabel?) -> Int {
        switch label {
        case .red: 1
        case .yellow: 2
        case .green: 3
        case .blue: 4
        case .purple: 5
        case .none: 0
        }
    }
}

private extension ColorLabel {
    var xmpLabelName: String {
        switch self {
        case .red: "Red"
        case .yellow: "Yellow"
        case .green: "Green"
        case .blue: "Blue"
        case .purple: "Purple"
        }
    }

    init?(displayName: String) {
        let normalized = displayName.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        switch normalized {
        case "red", "红色":
            self = .red
        case "yellow", "黄色":
            self = .yellow
        case "green", "绿色":
            self = .green
        case "blue", "蓝色":
            self = .blue
        case "purple", "紫色":
            self = .purple
        default:
            self.init(rawValue: normalized)
        }
    }
}

private extension String {
    var xmlEscaped: String {
        replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&apos;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
    }
}
